### PR TITLE
Fix _sync_dist deadlock for dist_reduce_fx=None list states

### PR DESCRIPTION
**Pull request by vedjaw**
## What does this PR do?

Fixes #3336

When a metric uses `dist_reduce_fx=None` list states (e.g. `MeanAveragePrecision`, retrieval metrics) and different ranks call `update()` a different number of times, `_sync_dist` deadlocks on `compute()` because `apply_to_collection` issues a different number of `all_gather` calls per rank.

This extends the empty-list fix from #2468 to `None`-reduction list states by:

1. Exchanging list lengths across ranks via `all_gather`
2. Exchanging tensor ndim / trailing shape metadata via `all_reduce` (all ranks participate, even when no padding is needed locally)
3. Padding shorter lists with ndim-matched empty tensors `(0, *tail_shape)` so every rank performs the same number of collectives
4. Adding barriers before/after equalization so no rank enters `apply_to_collection` early

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: src/torchmetrics/metric.py [hunk 1]
```diff
@@ -498,7 +498,61 @@ def _reduce_states(self, incoming_state: dict[str, Any]) -> None:
...
+    def _make_empty_pad_tensor(self, ref: Optional[Tensor], ndim: int, tail_shape_vec: Tensor) -> Tensor:
+        """Create an empty tensor with the same trailing dimensions as ``ref`` for list-state padding."""
+        if ref is not None:
+            if ref.ndim == 0:
+                return torch.zeros(0, device=self.device, dtype=ref.dtype)
+            return torch.zeros((0, *ref.shape[1:]), device=self.device, dtype=ref.dtype)
+        if ndim <= 1:
+            return torch.zeros(0, device=self.device, dtype=self.dtype)
+        pad_shape = [0] + [int(tail_shape_vec[i].item()) for i in range(ndim - 1)]
+        return torch.zeros(pad_shape, device=self.device, dtype=self.dtype)
```
[the remark this answers is not in the document]
Comment (vedjaw) [reply src/torchmetrics/metric.py]: Dtype half fixed in 120b99f. Ranks now agree on the element dtype over a small integer code before padding, so a rank with no elements pads with the real element dtype instead of `self.dtype`. Added a long-dtype regression test.

The 0-d half is real and I have **not** fixed it — flagging rather than guessing. `gather_all_tensors` branches to `_simple_gather_all_tensors` at `result.ndim == 0`, so padding a scalar list with a 1-D `zeros(0)` puts that index on the other branch and reintroduces the mismatch. The catch is that a 0-d tensor always holds exactly one element, so a shape-matching pad necessarily contributes a real value that has to be trimmed after the sync using the gathered lengths — which means threading those lengths out of `_equalize_none_reduction_list` into the post-`apply_to_collection` stage.

That is a behavioural choice (deadlock on scalar list states, as today, versus pad-and-trim), and I did not want to land an untested change to collective ordering. Happy to implement pad-and-trim if that is the direction you want.

File: src/torchmetrics/metric.py [hunk 2]
```diff
@@ -498,7 +498,61 @@ def _reduce_states(self, incoming_state: dict[str, Any]) -> None:
...
+        return state
+
     def _sync_dist(self, dist_sync_fn: Callable = gather_all_tensors, process_group: Optional[Any] = None) -> None:
+        group = process_group or self.process_group
+        if jit_distributed_available():
+            torch.distributed.barrier(group=group)
+
+        input_dict = {attr: getattr(self, attr) for attr in self._reductions}
+
         input_dict = {attr: getattr(self, attr) for attr in self._reductions}
```
[the remark this answers is not in the document]
Comment (vedjaw) [reply src/torchmetrics/metric.py]: Fixed in 120b99f — removed the duplicated `input_dict` assignment.

File: src/torchmetrics/metric.py [hunk 3]
```diff
@@ -498,7 +498,61 @@ def _reduce_states(self, incoming_state: dict[str, Any]) -> None:
...
+
+    def _equalize_none_reduction_list(self, state: list, process_group: Optional[Any] = None) -> list:
+        """Pad ``dist_reduce_fx=None`` list states so all ranks perform the same number of collectives."""
+        if not isinstance(state, list) or not jit_distributed_available():
+            return state
+
+        group = process_group or self.process_group
+        world_size = torch.distributed.get_world_size(group)
+        local_len = len(state)
+
```
[the remark this answers is not in the document]
Comment (vedjaw) [reply src/torchmetrics/metric.py:547]: Fixed in 120b99f. `_equalize_none_reduction_list` now returns early unless every entry is a `Tensor`, so states like `MeanAveragePrecision.detection_mask` pass through untouched — as you note, `apply_to_collection` dispatches on `Tensor` so they run no collectives and need no equalizing. Added a test with a list state of non-tensor objects.

File: tests/unittests/bases/test_ddp.py [hunk 4]
```diff
@@ -342,3 +343,53 @@ def _test_sync_with_unequal_size_lists(rank):
...
+
+    full_state_update = True
+
+    def __init__(self, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self.add_state("items", default=[], dist_reduce_fx=None)
+
+    def update(self, x: torch.Tensor) -> None:
+        self.items.append(x)
+
```
[the remark this answers is not in the document]
Comment (vedjaw) [reply tests/unittests/bases/test_ddp.py:359]: Fixed in 120b99f — added `_test_sync_none_reduction_non_default_dtype`, which updates with `torch.long` tensors on one rank only and asserts both the value and that the synced dtype stays `torch.long`. That is the case that would have caught the padding dtype bug.

**Comment by vedjaw** [comment]
Pushed 51eec91, which fixes a deadlock I introduced in the previous commit.

The non-tensor guard I added ran *before* any collective, so ranks could disagree on how many to run: a rank holding an empty list sees a vacuously all-tensor list and went on to the `all_gather` and `all_reduce`s, while a rank holding non-tensors returned immediately. That is the same mismatched-collective hang this method exists to prevent, and `_test_sync_none_reduction_non_tensor_list` sets it up exactly — rank 1 holds a tuple while rank 0 holds nothing.

The check now sits after the length `all_gather`, which every rank always reaches, and is agreed with a MIN `all_reduce` so any rank holding a non-tensor makes every rank skip. `max_len` was already safe to branch on since it comes from that `all_gather`.

Worth noting for the record: the `ubuntu24.04-cuda12.6.3-py3.12-torch2.8` job on the previous head ran for ~1h58m before failing while the other CUDA job passed, which is consistent with that hang rather than an assertion failure.
